proxy/src/connection.rs

```diff
@@ -252,24 +262,47 @@ impl Future for Connecting {
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
         loop {
-            *self = match self {
-                Connecting::Plaintext { connect, tls } => {
+            self.state = match &mut self.state {
+                ConnectingState::Plaintext { connect, tls } => {
+                    trace!("Connecting: state=plaintext; tls={:?};",tls);
                     let plaintext_stream = try_ready!(connect.poll());
                     set_nodelay_or_warn(&plaintext_stream);
                     match tls.take().expect("Polled after ready") {
                         Conditional::Some(config) => {
-                            let upgrade_to_tls = tls::Connection::connect(
+                            trace!("plaintext connection established; trying to upgrade");
+                            let upgrade = tls::Connection::connect(
                                 plaintext_stream, &config.identity, config.config);
-                            Connecting::UpgradeToTls(upgrade_to_tls)
+                            ConnectingState::UpgradeToTls(upgrade)
                         },
                         Conditional::None(why) => {
+                            trace!("plaintext connection established; no TLS ({:?})", why);
                             return Ok(Async::Ready(Connection::plain(plaintext_stream, why)));
                         },
                     }
                 },
-                Connecting::UpgradeToTls(upgrading) => {
-                    let tls_stream = try_ready!(upgrading.poll());
-                    return Ok(Async::Ready(Connection::tls(BoxedIo::new(tls_stream))));
+                ConnectingState::UpgradeToTls(upgrade) => {
+                    match upgrade.poll() {
+                        Ok(Async::NotReady) => return Ok(Async::NotReady),
+                        Ok(Async::Ready(tls_stream)) => {
+                            let conn = Connection::tls(BoxedIo::new(tls_stream));
+                            return Ok(Async::Ready(conn));
+                        },
+                        Err(e) => {
+                            warn!(
```

Comment:
Not a blocker: `warn!` is too high of a logging level for this because we expect this to happen a lot initially due to the delay in distributing the certificate secrets. I suggest `debug!`.